server/services/csvservices.go

```diff
@@ -13,18 +14,21 @@ import (
 
 // func to read csv files from URL
 func ReadCSVFromUrl(url string) ([][]string, error) {
+	var noCSV = errors.New("URL has no CSV.")
+	var invalidURL = errors.New("Invalid URL.")
+
 	resp, err := http.Get(url)
 	if err != nil {
```

Comment:
Added the following that better reflects the errors

```var error4xx = errors.New("invalid URL")
	var error5xx = errors.New("server error")
	var noCSV = errors.New("nothing to read at this URL")

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != 200 {
		if resp.StatusCode / 10 / 10 == 4 {
			return nil, error4xx
		} else if resp.StatusCode / 10 / 10 == 5 {
			return nil, error5xx
		}
	}
	if resp == nil {
		return nil, noCSV
	}
	defer resp.Body.Close()```

Reply:
This is not ideal, you need to check specific error code. Please understand what error codes may be thrown by this http.Get() method.

Reply:
How is this?

```
resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	statusCode := resp.StatusCode
	if statusCode != 200 {
		if statusCode == 401 || statusCode == 402 || statusCode == 403 || statusCode == 451 {
			return nil, errors.New("unable to gain access to website")
		} else if statusCode == 408 {
			return nil, errors.New("request timed out")
		} else if statusCode == 414 || statusCode == 431 {
			return nil, errors.New("URI too long")
		} else if statusCode == 429 {
			return nil, errors.New("too many requests sent at on time")
		} else if statusCode == 500 {
			return nil, errors.New("internal server error")
		} else if statusCode == 503 {
			return nil, errors.New("bad gateway")
		} else if statusCode == 504 {
			return nil, errors.New("gateway timeout")
		} else if statusCode == 507 {
			return nil, errors.New("server has insufficient storage to complete request")
		} else if statusCode == 508 {
			return nil, errors.New("infinite loop detected and server terminated process")
		} else if statusCode == 511 {
			return nil, errors.New("network authentication required to gain network access")
		} else {
			return nil, errors.New(fmt.Sprintf("unable to reach website with status code %d", resp.StatusCode))
		}
	}
	
	noCSV := errors.New("nothing to read at this URL")
	if resp == nil {
		return nil, noCSV
	}
```

Reply:
Refactor out as a method so that you can use it for any API call.
Instead of using the exact numbers, use the constants defined in go/http. Search for HTTP status codes and you can find all the constants there. https://golang.org/pkg/net/http/?m=all